app/javascript/packages/password-strength/password-strength-element.ts

```diff
@@ -0,0 +1,182 @@
+import zxcvbn from 'zxcvbn';
+import { t } from '@18f/identity-i18n';
+import type { ZXCVBNResult, ZXCVBNScore } from 'zxcvbn';
+
+const MINIMUM_STRENGTH: ZXCVBNScore = 3;
+
+const snakeCase = (string: string): string =>
+  string.replace(/[ -]/g, '_').replace(/\W/g, '').toLowerCase();
+
+class PasswordStrengthElement extends HTMLElement {
+  connectedCallback() {
+    this.input.addEventListener('input', () => this.#handleValueChange());
+  }
+
+  get strength(): HTMLElement {
+    return this.querySelector('.password-strength__strength')!;
+  }
+
+  get feedback(): HTMLElement {
+    return this.querySelector('.password-strength__feedback')!;
+  }
+
+  get input(): HTMLInputElement {
+    return this.ownerDocument.getElementById(this.getAttribute('input-id')!) as HTMLInputElement;
+  }
+
+  get minimumLength(): number {
+    return Number(this.getAttribute('minimum-length')!);
+  }
+
+  get forbiddenPasswords(): string[] {
+    return JSON.parse(this.getAttribute('forbidden-passwords')!);
+  }
+
+  /**
+   * Returns a normalized score on zxcvbn's scale. Notably, this artificially lowers a score if it
+   * does not meet the minimum length requires, to avoid confusion where an invalid value would
+   * display as being a great password.
+   *
+   * @param result zxcvbn result
+   *
+   * @return Normalized zxcvbn score
+   */
+  #getNormalizedScore(result: ZXCVBNResult): ZXCVBNScore {
+    const { score } = result;
+
+    if (this.input.value.length < this.minimumLength && score >= MINIMUM_STRENGTH) {
+      return Math.max(MINIMUM_STRENGTH - 1, 0) as ZXCVBNScore;
+    }
+
+    return score;
+  }
+
+  /**
+   * Returns true if the input's value is considered valid for submission, or false otherwise.
+   *
+   * @param result zxcvbn result
+   *
+   * @return Whether the input's value is valid for submission
+   */
+  #isValid(result: ZXCVBNResult): boolean {
+    return result.score >= MINIMUM_STRENGTH && this.input.value.length >= this.minimumLength;
+  }
```

Comment:
put the conditions in the same order for easier comparison?

```suggestion
  #getNormalizedScore(result: ZXCVBNResult): ZXCVBNScore {
    const { score } = result;

    if (score >= MINIMUM_STRENGTH && this.input.value.length < this.minimumLength) {
      return Math.max(MINIMUM_STRENGTH - 1, 0) as ZXCVBNScore;
    }

    return score;
  }

  /**
   * Returns true if the input's value is considered valid for submission, or false otherwise.
   *
   * @param result zxcvbn result
   *
   * @return Whether the input's value is valid for submission
   */
  #isValid(result: ZXCVBNResult): boolean {
    return result.score >= MINIMUM_STRENGTH && this.input.value.length >= this.minimumLength;
  }
```

Reply:
Makes sense 👍